theano/gpuarray/opt.py

```diff
@@ -600,58 +905,90 @@ def local_gpuajoin_1(node):
 
 @register_opt('fast_compile')
 @op_lifter([tensor.Split])
-def local_gpua_split(node, context_name):
-    return GpuSplit(node.op.len_splits)
+@register_opt2([tensor.Split], 'fast_compile')
+def local_gpua_split(op, context_name, inputs, outputs):
+    # TODO use props
+    return GpuSplit(op.len_splits)
 
 
 @register_opt('fast_compile')
 @op_lifter([tensor.Subtensor])
-def local_gpua_subtensor(node, context_name):
-    x = node.inputs[0]
+def local_gpua_subtensor(op, context_name, inputs, outputs):
+    x = inputs[0]
     if (x.owner and isinstance(x.owner.op, HostFromGpu)):
         gpu_x = x.owner.inputs[0]
         if (gpu_x.owner and
                 isinstance(gpu_x.owner.op, GpuFromHost) and
                 # And it is a shared var or an input of the graph.
                 not gpu_x.owner.inputs[0].owner):
-            if len(x.clients) == 1:
-                if any([n == 'output' or any([isinstance(v.type, GpuArrayType)
-                                              for v in n.inputs + n.outputs])
-                        for n, _ in node.outputs[0].clients]):
-                    return
-                else:
-                    return [host_from_gpu(gpu_x.owner.op(node.outputs[0]))]
-
-    return GpuSubtensor(node.op.idx_list)
+            if len(x.clients) == 1 and len(outputs[0].clients) == 1:
+                return
+    # Here is the condition for the GraphToGPU opt. inputs is the
+    # inputs we want to use for the new node
+    if (x.owner and isinstance(x.owner.op, GpuFromHost)):
+        cpu_x = x.owner.inputs[0]
+        # And it is a shared var or an input of the graph.
+        # and is used by only 1 node.
+        # x is in the new graph, so we can't tests its number of clients.
+        if not cpu_x.owner and len(cpu_x.clients) == 1:
+            return
+    return GpuSubtensor(op.idx_list)
+
+
+@register_opt2([tensor.Subtensor], 'fast_compile')
+def local_gpua_subtensor_graph(op, context_name, inputs, outputs):
+    # We need different code as the condition is different as inputs
+    # aren't the same.
+    x = inputs[0]
+    # We don't want to move the subtensor to the GPU if the inputs is
+    # on the CPU and the only client of the CPU node is this
+    # subtensor. This allow to have a smaller transfer.
+
+    if (x.owner and isinstance(x.owner.op, GpuFromHost)):
+        cpu_x = x.owner.inputs[0]
+        # And it is a shared var or an input of the graph.
+        # and is used by only 1 node.
+        # x is in the new graph, so we can't tests its number of clients.
+        if not cpu_x.owner and len(cpu_x.clients) == 1:
+            c = outputs[0].clients
+            # If the subtensor have only 1 client, do it on the CPU.
+            # We let the other optimization to take care to move the
+            # next node or not.
+            if len(c) == 1:
+                return
+    return GpuSubtensor(op.idx_list)
```

Comment:
It may be correct, but I would really prefer there to be a test that details exactly what we expect to happen in various cases, and making sure it is actually happening.
Right now, it looks like something that works, but we don't know how.